docs/source/generics.rst

```diff
@@ -0,0 +1,984 @@
+Generics
+========
+
+You may have seen type hints like ``list[str]`` or ``dict[str, int]`` in Python
+code. These types are interesting in that they are parametrised by other types!
+A ``list[str]`` isn't just a list, it's a list of strings. Types with type
+parameters like this are called *generic types*.
+
+You can define your own generic classes that take type parameters, similar to
+built-in types such as ``list[X]``. Note that such user-defined generics are a
+moderately advanced feature and you can get far without ever using them.
+
+.. _generic-classes:
+
+Defining generic classes
+************************
+
+Here is a very simple generic class that represents a stack:
+
+.. code-block:: python
+
+   from typing import TypeVar, Generic
+
+   T = TypeVar('T')
+
+   class Stack(Generic[T]):
+       def __init__(self) -> None:
+           # Create an empty list with items of type T
+           self.items: list[T] = []
+
+       def push(self, item: T) -> None:
+           self.items.append(item)
+
+       def pop(self) -> T:
+           return self.items.pop()
+
+       def empty(self) -> bool:
+           return not self.items
+
+The ``Stack`` class can be used to represent a stack of any type:
+``Stack[int]``, ``Stack[tuple[int, str]]``, etc.
+
+Using ``Stack`` is similar to built-in container types, like ``list``:
+
+.. code-block:: python
+
+   # Construct an empty Stack[int] instance
+   stack = Stack[int]()
+   stack.push(2)
+   stack.pop() + 1
+   stack.push('x')  # error: Argument 1 to "push" of "Stack" has incompatible type "str"; expected "int"
+
+When creating instances of generic classes, the type parameter can usually be
```

Comment:
Nit: s/type parameter/type argument/ here and below.

Reply:
Is this right? I meant in the case where there is no argument

Reply:
Yes, I think that is right. Type parameters are the type variables used in the definition of a generic class. Type arguments are the types a generic class is specialized with. So, below `T` is a type parameter of `Box` and `int` in `Box[int]()` is a type argument.